quodlibet/util/matcher.py

```diff
@@ -0,0 +1,382 @@
+        # would be somewhat unpredictable.
+        #
+        # For example, a b_item might be undefeatable with a similarity of 0.52. If we
+        # wouldn't normalize it, it would not pass a minimum of 0.6, even if it could
+        # easily achieve that with the remaining weights.
+        if cont_attr_idx < len(self._attr_with_weight):
+            cur_total_weight = 1 - self._weight_left[cont_attr_idx]
+            similarity /= cur_total_weight
+
+        return similarity < self.minimum_similarity_ratio
+
+    def _handle_conflicts_if_any(self, a1_match_data):
+        while True:
+            best_b_idx = a1_match_data.best_b_idx
+            if best_b_idx is None:
+                # a1 has no matches left or could not find a match
+                return
+
+            a2_match_data = self._b_idx_to_a_match_data[best_b_idx]
+            if a2_match_data is None:
+                self._b_idx_to_a_match_data[best_b_idx] = a1_match_data
+                return
+
+            # We have found a conflict and will now solve it.
+            a1_match_data = self._get_worse_match_data(a1_match_data, a2_match_data)
+
+    def _get_worse_match_data(self, a1_match_data, a2_match_data):
+        # b is matched to a previous a (a2), so we have to find a better match
+        self._finish_similarity_measures(a1_match_data)
+        self._finish_similarity_measures(a2_match_data)
+
+        # As the index of a1 will almost always be larger than a2, we use <=
+        # here, since in case they're equal in terms of similarity, we want to
+        # give some weight to the current order of b (index 0 preferred to 1).
+        if a1_match_data.best_b_similarity <= a2_match_data.best_b_similarity:
+            a1_match_data.replace_best()
+
+            # still need to find a better match for a1
+            return a1_match_data
+        else:
+            # a1 is better so replace a2 in map
+            best_b_idx = a1_match_data.best_b_idx
+            self._b_idx_to_a_match_data[best_b_idx] = a1_match_data
+
+            a2_match_data.replace_best()
+
+            # we need to find a new match for a2
+            return a2_match_data
+
+    def _finish_similarity_measures(self, a_match_data):
+        """Only if this was called, is match data full measured and only then
+        can match_data.replace_best() be called."""
+        if a_match_data.is_fully_measured:
+            return
+
+        # figure out total similarity (without stopping) if we didn't
+        self._measure_similarity_to_find_best_b_match(a_match_data)
+        a_match_data.set_to_fully_measured()
+
+    def _measure_similarity_to_find_best_b_match(self, a_match_data):
+        continue_attr_idx = a_match_data.continue_attr_index
+        can_stop = continue_attr_idx == 0 and not self.should_go_through_every_attribute
+        attr_size = len(self._attr_with_weight)
+
+        for attr_idx in range(continue_attr_idx, attr_size):
```

Comment:
why not just iterate over `self._attr_with_weight` then (or use `enumerate`)?

Reply:
The `continue_attr_idx` is not always `0` and we don't need the weight here.